streams/src/test/java/org/apache/kafka/streams/tests/StreamsBrokerDownResilienceTest.java

```diff
@@ -0,0 +1,166 @@
+
+        streams.setUncaughtExceptionHandler(new Thread.UncaughtExceptionHandler() {
+            @Override
+            public void uncaughtException(final Thread t, final Throwable e) {
+                System.err.println("FATAL: An unexpected exception " + e);
+                System.err.flush();
+                streams.close(30, TimeUnit.SECONDS);
+            }
+        });
+        System.out.println("start Kafka Streams");
+        streams.start();
+
+        System.out.println("Waiting during broker downtime starting now " + new Date());
+        try {
+            Thread.sleep(brokerDownTime + 10000L);
+        } catch (InterruptedException e) {
+            System.err.println("Pause during broker down time interrupted");
+            Thread.currentThread().interrupt();
+        }
+
+        System.out.println("Back after broker time and see if still viable by sending messages now " + new Date());
+
+        System.out.println("send data");
+        final Properties producerProperties = new Properties();
+        producerProperties.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, kafka);
+        producerProperties.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class);
+        producerProperties.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, StringSerializer.class);
+
+        final KafkaProducer<String, String> producer = new KafkaProducer<>(producerProperties);
+        producer.send(new ProducerRecord<>(SOURCE_TOPIC_1, "key", "value"));
+
+
+        System.out.println("wait for result");
+        loopUntilRecordReceived(kafka);
+
+        System.out.println("close Kafka Streams");
+        producer.close();
+        streams.close();
+    }
+
+
+    private static void loopUntilRecordReceived(final String kafka) {
+        final Properties consumerProperties = new Properties();
+        consumerProperties.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, kafka);
+        consumerProperties.put(ConsumerConfig.GROUP_ID_CONFIG, "stream-resilience-consumer");
+        consumerProperties.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
+        consumerProperties.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class);
+        consumerProperties.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class);
+
+        final KafkaConsumer<String, String> consumer = new KafkaConsumer<>(consumerProperties);
+        consumer.subscribe(Collections.singletonList(SINK_TOPIC));
+        while (true) {
+            final ConsumerRecords<String, String> records = consumer.poll(100);
+            for (final ConsumerRecord<String, String> record : records) {
+                if (record.key().equals("key") && record.value().equals("value")) {
+                    System.out.println("Received expected results");
+                    System.out.flush();
+                    consumer.close();
+                    return;
+                }
```

Comment:
What for the `else` part? `else` should never happen, right? Should we throw or write an error msg? Maybe move `return` out of `if` part?